src/main/java/org/broadinstitute/hellbender/utils/hdf5/HDF5PoNCreator.java

```diff
@@ -0,0 +1,712 @@
+    static double subtractBGSCenter(final ReadCountCollection readCounts, final Logger logger) {
+        final RealMatrix counts = readCounts.counts();
+        final int columnCount = counts.getColumnDimension();
+        final double[] columnMedians = new double[columnCount];
+        final double[][] data = counts.getData();
+        final Median medianCalculator = new Median();
+        final double[] columnData = new double[data.length];
+        for (int i = 0; i < columnCount; i++) {
+            for (int j = 0; j < data.length; j++) {
+                columnData[j] = data[j][i];
+            }
+            columnMedians[i] = medianCalculator.evaluate(columnData);
+        }
+        final double medianOfMedians = medianCalculator.evaluate(columnMedians);
+        counts.walkInOptimizedOrder(new DefaultRealMatrixChangingVisitor() {
+            @Override
+            public double visit(final int row, final int column, final double value) {
+                return value - medianOfMedians;
+            }
+        });
+        logger.info(String.format("Counts centered around the BGS center %.2f", medianOfMedians));
+        return medianOfMedians;
+    }
+
+    /**
+     * Final pre-panel normalization that consists on dividing all counts by the median of
+     * its column and log it with base 2.
+     * <p>
+     *     The normalization occurs in-place.
+     * </p>
+     *
+     * @param readCounts the input counts to normalize.
+     */
+    @VisibleForTesting
+    static void normalizeAndLogReadCounts(final ReadCountCollection readCounts, final Logger logger) {
+        final RealMatrix counts = readCounts.counts();
+        final int columnCount = counts.getColumnDimension();
+        final Median medianCalculator = new Median();
+        for (int i = 0; i < columnCount; i++) {
+            final double[] colCounts = counts.getColumn(i);
+            final double median = medianCalculator.evaluate(colCounts);
+            final double invMedian = 1.0 / median;
+            for (int j = 0; j < colCounts.length; j++) {
+                colCounts[j] = Math.log(Math.max(EPSILON, colCounts[j] * invMedian)) * INV_LN_2;
+            }
+            counts.setColumn(i, colCounts);
+        }
+        logger.info("Counts normalized by the column median and changed into bits.");
+    }
+
+
+    /**
+     * Impute zero counts to the median of non-zero values in the enclosing target row.
+     *
+     * <p>The imputation is done in-place, thus the input matrix is well be modified as a result of this call.</p>
+     *
+     * @param readCounts the input and output read-count matrix.
+     */
+    @VisibleForTesting
+    static void imputeZerosCounts(final ReadCountCollection readCounts, final Logger logger) {
+
+        final RealMatrix counts = readCounts.counts();
+        final int targetCount = counts.getRowDimension();
+        final int columnCount = counts.getColumnDimension();
+
+        // Declared once outside loop to reuse across rows:
```

Comment:
IMO there's premature optimization going on.  At the cost of doing two passes instead of one, we could replace this with:
```
final double[] medians = IntStream(0, targetCount).map(t -> counts.getRow(t).stream().filter(c -> c != 0.0).median());
counts.walkInOptimizedOrder(new DefaultRealMatrixChangingVisitor() {
            @Override
            public double visit(final int row, final int column, final double value) { return value != 0 ? value : medians[row]; }
        });
```